core/src/layers/foyer.rs

```
@@ -0,0 +1,191 @@
+use crate::raw::*;
+use crate::*;
+use chrono::{DateTime, Utc};
+use foyer::{Code, CodeError, HybridCache};
+use serde::{Deserialize, Serialize};
+use std::sync::Arc;
+
+pub struct FoyerLayer {
+    cache: Arc<HybridCache<CacheKey, CacheValue>>,
+}
+
+impl FoyerLayer {
+    pub async fn new(cache: HybridCache<CacheKey, CacheValue>) -> Result<Self> {
+        Ok(Self {
+            cache: Arc::new(cache),
+        })
+    }
+}
+
+impl<A: Access> Layer<A> for FoyerLayer {
+    type LayeredAccess = CacheAccessor<A>;
+
+    fn layer(&self, inner: A) -> Self::LayeredAccess {
+        CacheAccessor {
+            inner,
+            cache: Arc::clone(&self.cache),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Hash, Eq, PartialEq, Serialize, Deserialize)]
+pub struct CacheKey {
+    path: String,
+    range: BytesRange,
+    if_match: Option<String>,
+    if_none_match: Option<String>,
+    if_modified_since: Option<DateTime<Utc>>,
+    if_unmodified_since: Option<DateTime<Utc>>,
+    version: Option<String>,
+}
+
+impl CacheKey {
+    fn new(path: &str, args: &OpRead) -> CacheKey {
+        CacheKey {
+            path: path.to_string(),
+            range: args.range(),
+            if_match: args.if_match().map(ToString::to_string),
+            if_none_match: args.if_none_match().map(ToString::to_string),
+            if_modified_since: args.if_modified_since(),
+            if_unmodified_since: args.if_unmodified_since(),
+            version: args.version().map(ToString::to_string),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct CacheValue(Buffer);
+
+impl Code for CacheValue {
```

I wonder @MrCroxx how should we take care of `CacheKey` and `CacheValue` evolution throught time. Should we take care of breaking changes such including a new mandatory field in the `CacheKey` struct? 

For example, if we add a new mandatory field to `CacheKey`, previous version's keys in disk cache will fail to deserialize? How should we do things here? We must always do back-compatible changes to both `CacheKey` and `CacheValue`?

---

There is no deserialize for `CacheKey`. The cache just try to read the key we input. And `CacheValue` itself is a plain `Buffer` which contains the raw content.

I think we don't need to concern about the back-compatible.

---

I think deserialization is done for the cache key. 

https://github.com/foyer-rs/foyer/blob/efa47f7c70ed70f0970d2b04ffc8e6422b5c78c9/foyer-storage/src/serde.rs#L178

---

> I think deserialization is done for the cache key.

Good point!

cc @MrCroxx, what happens if the cache key changes? Will Foyer silently ignore it and remove the stale cache value, or will we encounter an error?